contracts/distribution/subdomains/ZNSSubdomainRegistrar.sol

```diff
@@ -0,0 +1,229 @@
+    IZNSRegistrar public rootRegistrar;
+
+    // TODO sub: make better name AND for the setter function !
+    mapping(bytes32 domainHash => DistributionConfig) public distrConfigs;
+
+    mapping(bytes32 domainHash =>
+        mapping(address registrant => bool allowed)
+    ) public distributionWhitelist;
+
+    modifier onlyOwnerOperatorOrRegistrar(bytes32 domainHash) {
+        require(
+            registry.isOwnerOrOperator(domainHash, msg.sender)
+            || accessController.isRegistrar(msg.sender),
+            "ZNSSubdomainRegistrar: Not authorized"
+        );
+        _;
+    }
+
+    // TODO sub: proxy ??
+    constructor(
+        address _accessController,
+        address _registry,
+        address _rootRegistrar
+    ) {
+        _setAccessController(_accessController);
+        setRegistry(_registry);
+        setRootRegistrar(_rootRegistrar);
+    }
+
+    function registerSubdomain(
+        bytes32 parentHash,
+        string calldata label,
+        address domainAddress,
+        DistributionConfig calldata distrConfig
+    ) external override returns (bytes32) {
+        // TODO sub: make the order of ops better
+        DistributionConfig memory parentConfig = distrConfigs[parentHash];
+        require(
+            parentConfig.accessType != AccessType.LOCKED
+                || registry.isOwnerOrOperator(parentHash, msg.sender),
+            "ZNSSubdomainRegistrar: Parent domain's distribution is locked"
+        );
+
+        if (parentConfig.accessType == AccessType.WHITELIST) {
+            require(
+                distributionWhitelist[parentHash][msg.sender],
+                "ZNSSubdomainRegistrar: Sender is not whitelisted"
+            );
+        }
+
+        bytes32 subdomainHash = hashWithParent(parentHash, label);
+
+        require(
+            !registry.exists(subdomainHash),
+            "ZNSSubdomainRegistrar: Subdomain already exists"
+        );
+
+        uint256 price;
+        uint256 fee;
+        // TODO sub: can we make this abstract switching better ??
+        if (parentConfig.pricingContract.feeEnforced()) {
+            (price, fee) = AZNSPricingWithFee(address(parentConfig.pricingContract)).getPriceAndFee(
+                parentHash,
+                label
+            );
+        } else {
+            price = parentConfig.pricingContract.getPrice(parentHash, label);
+        }
+
+        parentConfig.paymentContract.processPayment(
+            parentHash,
+            subdomainHash,
+            msg.sender,
+            price,
+            fee
+        );
+
+        rootRegistrar.coreRegister(
+            parentHash,
+            subdomainHash,
+            label,
+            msg.sender,
+            domainAddress
+        );
+
+        if (address(distrConfig.pricingContract) != address(0)
+            && address(distrConfig.paymentContract) != address(0)) {
+            setDistributionConfigForDomain(subdomainHash, distrConfig);
+        }
```

Comment:
Misconfigured where and how exactly?
This is basically an option for the user to either set the config upon registration or set it later.

Reply:
As in, the contracts provided don't adhere to the erc165 interface.
So in the places here where you're calling things like parentConfig.accessType, these will break if the contract doesn't have this member.

But it may just be fine, as you've said it's up to the module to work properly.